api/handlers/jwt.go

```diff
@@ -11,56 +11,82 @@ import (
 	"strings"
 	"time"
 
+	"github.com/labstack/echo"
 	"github.com/spf13/viper"
 	jose "gopkg.in/square/go-jose.v2"
 )
 
-type idToken struct {
-	Issuer          string                 `json:"iss"`
-	Subject         string                 `json:"sub"`
-	Audience        audience               `json:"aud"`
-	AuthorizedParty string                 `json:"azp"`
-	Expiry          jsonTime               `json:"exp"`
-	IssuedAt        jsonTime               `json:"iat"`
-	Nonce           string                 `json:"nonce"`
-	AtHash          string                 `json:"at_hash"`
-	ClaimNames      map[string]string      `json:"_claim_names"`
-	ClaimSources    map[string]claimSource `json:"_claim_sources"`
+// IDToken is the struct that holds all information about a JWT token
+type IDToken struct {
+	Issuer            string                 `json:"iss"`
+	Subject           string                 `json:"sub"`
+	Audience          audience               `json:"aud"`
+	AuthorizedParty   string                 `json:"azp"`
+	Expiry            jsonTime               `json:"exp"`
+	IssuedAt          jsonTime               `json:"iat"`
+	Nonce             string                 `json:"nonce"`
+	AtHash            string                 `json:"at_hash"`
+	Name              string                 `json:"name"`
+	PreferredUsername string                 `json:"preferred_username"`
+	GivenName         string                 `json:"given_name"`
+	FamilyName        string                 `json:"family_name"`
+	MiddleName        string                 `json:"middle_name"`
+	Nickname          string                 `json:"nickname"`
+	PhoneNumber       string                 `json:"phone_number"`
+	Email             string                 `json:"email"`
+	ClaimNames        map[string]string      `json:"_claim_names"`
+	ClaimSources      map[string]claimSource `json:"_claim_sources"`
 }
+
 type claimSource struct {
 	Endpoint    string `json:"endpoint"`
 	AccessToken string `json:"access_token"`
 }
 type jsonTime time.Time
 type audience []string
 
-// ValidateJWT validates JWT based on audience, expiration, signature and issuer
-func ValidateJWT(jwt string, config viper.Viper) error {
+// ValidateJWT validates JWT based on audience, expiration, signature and issuer and returns a valid JWT token if it succeeds
+func ValidateJWT(c echo.Context, config viper.Viper) (IDToken, error) {
 	var err error
-	err = verifyAudience(jwt, config.GetString("oidc_audience"))
+	token := IDToken{}
+	authorizationHeader := c.Request().Header.Get("Authorization")
+	if len(authorizationHeader) == 0 {
+		return token, fmt.Errorf("JWT: %s", err.Error())
+	}
+	jwtSlice := strings.Split(authorizationHeader, "JWT")
+	if len(jwtSlice) != 1 {
+		return token, fmt.Errorf("JWT: %s", err.Error())
+	}
+	jwt := jwtSlice[0]
+
+	token, err = parseIDToken(jwt)
 	if err != nil {
-		return fmt.Errorf(err.Error())
+		return token, fmt.Errorf(err.Error())
```

Comment:
If error exists, this return can be:
```
return nil, fmt.Errorf(err.Error())
```

Reply:
I checked this, but this function returns a `IDToken` type. It is not possible returns a `nil`.